src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1488 @@
+                codebook_preds = codebook_preds.to(torch.int32)
+                input_buffer[:, rel_start_fill_idx:, n_inner] = codebook_preds
+                del logits, codebook_preds
+
+            # transfer into fine_input
+            for n_inner in range(n_coarse, fine_generation_config.n_fine_codebooks):
+                fine_input[
+                    :, start_fill_idx : start_fill_idx + (max_fine_input_length - rel_start_fill_idx), n_inner
+                ] = input_buffer[:, rel_start_fill_idx:, n_inner]
+            del input_buffer
+
+        fine_input = fine_input.transpose(1, 2)[:, :, n_history:]
+        if n_remove_from_end > 0:
+            fine_input = fine_input[:, :, :-n_remove_from_end]
+
+        if fine_input.shape[-1] != coarse_output.shape[-2]:
+            raise ValueError("input and output should have the same seq_len")
+
+        return fine_input
+
+
+@add_start_docstrings(
+    """
+    The full Bark model, a text-to-speech model composed of 4 sub-models:
+    - [`BarkSemanticModel`] (also referred to as the 'text' model): a causal auto-regressive transformer model that
+      takes
+    as input tokenized text, and predicts semantic text tokens that capture the meaning of the text.
+    - [`BarkCoarseModel`] (also refered to as the 'coarse acoustics' model), also a causal autoregressive transformer,
+    that takes into input the results of the last model. It aims at regressing the first two audio codebooks necessary
+    to `encodec`.
+    - [`BarkFineModel`] (the 'fine acoustics' model), this time a non-causal autoencoder transformer, which iteratively
+    predicts the last codebooks based on the sum of the previous codebooks embeddings.
+    - having predicted all the codebook channels from the [`EncodecModel`], Bark uses it to decode the output audio
+      array.
+
+    It should be noted that each of the first three modules can support conditional speaker embeddings to condition the
+    output sound according to specific predefined voice.
+    """,
+    BARK_START_DOCSTRING,
+)
+class BarkModel(BarkPreTrainedModel):
+    config_class = BarkConfig
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.semantic = BarkSemanticModel(config.semantic_config)
+        self.coarse_acoustics = BarkCoarseModel(config.coarse_acoustics_config)
+        self.fine_acoustics = BarkFineModel(config.fine_acoustics_config)
+
+        self.codec_model = AutoModel.from_config(config.codec_config)
+
+        self.config = config
+
+    def codec_decode(self, fine_output):
+        """Turn quantized audio codes into audio array using encodec."""
+
+        fine_output = fine_output.transpose(0, 1)
+        emb = self.codec_model.quantizer.decode(fine_output)
+        out = self.codec_model.decoder(emb)
+        audio_arr = out.squeeze(1)  # squeeze the codebook dimension
+
+        return audio_arr
+
+    @torch.no_grad()
+    def generate_speech(
```

Comment:
Not a fan of all those `generate_xxx` methods really. Again would prefer to just have `generate`.

`ImageGPT` does not have `generate_images` for instance.

Reply:
Probably this would be best for consistency across the repo, i.e. every model should strive to generate with `model.generate`, rather than introducing its own API

If each of the sub-models has its own generation method, these can be called `generate` without any infringement on the rest of the code. The overall model can then have a `generate` method that just calls the sub-models, e.g.:

```python
def generate(...)
    semantic_ids = self.semantic_model.generate(...)

    coarse_ids = self.coarse_model.generate(semantic_ids, ...)
    
    ...
````

Reply:
Agreed, I'm not a fan of `generate_speech` in `SpeechT5` either but let it slip through review 😬

Reply:
Hi @sgugger and @sanchit-gandhi, thanks for the insightful review! I totally understand and support the need to stay consistent with transformers! 
I might have some questions regarding how to best test the sub-models though. 
At the moment, I test the classic `GenerationMixin.generate` method without too much difficulty by inheriting the tests from `GenerationTesterMixin` ([here](https://github.com/ylacombe/transformers/blob/2ce713e31c244105e18e76c46e8ddd84bf58e7ff/tests/models/bark/test_modeling_bark.py#L476) for example). 
If I modify the `generate` method of the sub-models, I can't test the `GenerationMixin.generate` of these sub-models as simply. Do you have any ideas of how I could simply test those without copying and modifying every tests of `GenerationTesterMixin` ?

Reply:
I don't understand the question: the tests are not testing your `generate_xxx` method but the base method which is irrelevant.

Reply:
Testing the base methods is still relevant IMO because the models tested are GPT2-like auto-regressive models and because the base methods are used in `generate_xxx`. Users might want to use `generate_xxx` with an underlying generation method different from the base generation method (multinomial sampling).

BTW `generate_xxx` methods are tested [here](https://github.com/ylacombe/transformers/blob/2ce713e31c244105e18e76c46e8ddd84bf58e7ff/tests/models/bark/test_modeling_bark.py#L801).

Reply:
Yes I could do that indeed. At the end of the day, I did something that allow to keep `GenerationMixinTester` of the base `generate`.